webapi/setaltsignaddr_test.go

```diff
@@ -167,107 +173,119 @@ func TestSetAltSignAddress(t *testing.T) {
-		wantCode:         http.StatusBadRequest,
+		name: "ticket can't vote",
+		addr: testAddr,
+		node: &testNode{
+			getRawTransaction: &dcrdtypes.TxRawResult{
+				Confirmations: 1000,
+			},
+			existsLiveTicket: false,
+		},
+		wantCode: http.StatusBadRequest,
 	}, {
-		name:                  "hist at max",
-		addr:                  testAddr,
+		name: "only one alt sign addr allowed",
+		addr: testAddr,
+		node: &testNode{
+			getRawTransaction: &dcrdtypes.TxRawResult{},
+			existsLiveTicket:  true,
+		},
 		isExistingAltSignAddr: true,
 		wantCode:              http.StatusBadRequest,
 	}}
 
 	for _, test := range tests {
-		ticketHash := randString(64, hexCharset)
-		req := &setAltSignAddrRequest{
-			Timestamp:      time.Now().Unix(),
-			TicketHash:     ticketHash,
-			TicketHex:      randString(504, hexCharset),
-			ParentHex:      randString(504, hexCharset),
-			AltSignAddress: test.addr,
-		}
-		reqSig := randString(504, hexCharset)
-		b, err := json.Marshal(req)
-		if err != nil {
-			t.Fatal(err)
-		}
-
-		if test.isExistingAltSignAddr {
-			data := &database.AltSignAddrData{
-				AltSignAddr: test.addr,
-				Req:         string(b),
-				ReqSig:      reqSig,
-				Resp:        string(randBytes(1000)),
-				RespSig:     randString(96, sigCharset),
+		t.Run(test.name, func(t *testing.T) {
+			ticketHash := randString(64, hexCharset)
+			req := &setAltSignAddrRequest{
+				Timestamp:      time.Now().Unix(),
+				TicketHash:     ticketHash,
+				TicketHex:      randString(504, hexCharset),
+				ParentHex:      randString(504, hexCharset),
+				AltSignAddress: test.addr,
+			}
+			reqSig := randString(504, hexCharset)
+			b, err := json.Marshal(req)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if test.isExistingAltSignAddr {
+				data := &database.AltSignAddrData{
+					AltSignAddr: test.addr,
+					Req:         string(b),
+					ReqSig:      reqSig,
+					Resp:        string(randBytes(1000)),
+					RespSig:     randString(96, sigCharset),
+				}
+				if err := api.db.InsertAltSignAddr(ticketHash, data); err != nil {
+					t.Fatalf("unable to insert alt sign addr: %v", err)
+				}
+			}
+
+			api.cfg.VspClosed = test.vspClosed
+
+			w := httptest.NewRecorder()
+			c, r := gin.CreateTestContext(w)
+
+			var dcrdErr error
+			if test.dcrdClientErr {
+				dcrdErr = errors.New("error")
+			}
+
+			handle := func(c *gin.Context) {
+				c.Set(dcrdKey, test.node)
+				c.Set(dcrdErrorKey, dcrdErr)
+				c.Set(requestBytesKey, b[test.deformReq:])
+				api.setAltSignAddr(c)
 			}
-			if err := api.db.InsertAltSignAddr(ticketHash, data); err != nil {
-				t.Fatalf("%q: unable to insert alt sign addr: %v", test.name, err)
+
+			r.POST("/", handle)
+
+			c.Request, err = http.NewRequest(http.MethodPost, "/", nil)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			c.Request.Header.Set("VSP-Client-Signature", reqSig)
+
+			r.ServeHTTP(w, c.Request)
+
+			if test.wantCode != w.Code {
+				t.Errorf("expected status %d, got %d", test.wantCode, w.Code)
```

Comment:
I guess I did this, but should be t.Fatalf, otherwise test may panic if this errors and test continues.